File: leetcode/medium/Q98_ValidateBinarySearchTree.py
import sys


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        if not root:
            return True
        return self.find(root,  -sys.maxsize - 1, sys.maxsize)

    def find(self, node: TreeNode, start: int, end: int) -> bool:
        if node.left is None and node.right is None:
            return True
        answer = True
        if node.left:
            if node.left.val >= node.val or node.left.val <= start:
                return False
            else:
                answer = answer and self.find(node.left, start, min(end, node.val))
        if node.right:
            if node.right.val <= node.val or node.right.val >= end:
                return False
            else:
                answer = answer and self.find(node.right, max(start, node.val), end)
        return answer

File: src/medium/Q98_ValidateBinarySearchTree.java
package medium;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class Q98_ValidateBinarySearchTree {
   
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
          this.val = val;
          this.left = left;
          this.right = right;
        }
    }
   
    class Solution {
        List<TreeNode> valList;
        public boolean isValidBST(TreeNode root) {
            valList = new ArrayList<>();
            traverse(root);
            int length = valList.size();
            for (int i = 0; i < length - 1; i++) {
                if(valList.get(i).val >= valList.get(i+1).val) return false;
            }
            return true;
        }
        public void traverse(TreeNode node){
            if(node.left != null) {
                traverse(node.left);
            }
            valList.add(node);
            if(node.right != null){
                traverse(node.right);
            }
        }
    }
}
